File: beecrowd/1947 - Rota do Taxista.rs
#![allow(dead_code)]
use std::{
    cmp::Reverse,
    collections::BinaryHeap,
    io::{self, Read, Write},
};

struct Scanner {
    buffer: Vec<u8>,
    index: usize,
}

impl Scanner {
    fn new() -> Self {
        let mut input = Vec::new();

        io::stdin().read_to_end(&mut input).unwrap();

        Self {
            buffer: input,
            index: 0,
        }
    }

    fn has_next(&mut self) -> bool {
        while self.index < self.buffer.len() && self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        self.index < self.buffer.len()
    }

    fn next<T: std::str::FromStr>(&mut self) -> T {
        if !self.has_next() {
            panic!("End Of File");
        }

        let start = self.index;

        while self.index < self.buffer.len() && !self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        std::str::from_utf8(&self.buffer[start..self.index])
            .unwrap()
            .parse::<T>()
            .ok()
            .unwrap()
    }
}

fn dijkstra(n: usize, adj: &[Vec<(usize, i32)>], src: usize, dest: usize) -> i32 {
    let mut dist = vec![i32::MAX; n];
    let mut pq = BinaryHeap::new();

    dist[src] = 0;
    pq.push(Reverse((0, src)));

    while let Some(Reverse((cost, v))) = pq.pop() {
        if cost != dist[v] {
            continue;
        } else if v == dest {
            return cost;
        }

        for &(u, w) in &adj[v] {
            let total_cost = cost + w;

            if dist[u] <= total_cost {
                continue;
            }

            dist[u] = total_cost;
            pq.push(Reverse((total_cost, u)));
        }
    }

    panic!("not found");
}

fn main() {
    let mut writer = io::BufWriter::with_capacity(1 << 20, io::stdout());

    let mut sc = Scanner::new();

    let n = sc.next::<usize>();
    let m = sc.next::<usize>();
    let k = sc.next::<usize>();

    let mut adj = vec![vec![]; n];

    for _ in 0..m {
        let u = sc.next::<usize>() - 1;
        let v = sc.next::<usize>() - 1;
        let w = sc.next::<i32>();

        adj[u].push((v, w));
        adj[v].push((u, w));
    }

    let mut travels = Vec::with_capacity(k + 1);

    for _ in 0..k {
        let src = sc.next::<usize>() - 1;
        let dest = sc.next::<usize>() - 1;
        travels.push((src, dest));
    }

    travels.push((0, 0));

    let mut pairwise_dist = vec![vec![0; k + 1]; k + 1];

    for (i, &pairfrom) in travels.iter().enumerate() {
        let (src1, dest1) = pairfrom;

        pairwise_dist[i][i] = dijkstra(n, &adj, src1, dest1);

        for (j, &pairto) in travels.iter().enumerate() {
            let (src2, _) = pairto;

            if i != j {
                pairwise_dist[i][j] = dijkstra(n, &adj, dest1, src2);
            }
        }
    }

    const INF: i32 = 1_000_000_000;
    let mut dp = vec![vec![INF; k]; 1 << k];

    for i in 0..k {
        dp[1 << i][i] = pairwise_dist[k][i] + pairwise_dist[i][i];
    }

    for i in 0..(1 << k) {
        for l in 0..k {
            if i & (1 << l) == 0 {
                continue;
            }

            for j in 0..k {
                if i & (1 << j) == 0 {
                    continue;
                }

                dp[i][j] = std::cmp::min(
                    dp[i ^ (1 << j)][l] + pairwise_dist[l][j] + pairwise_dist[j][j],
                    dp[i][j],
                );
            }
        }
    }

    let mut min_travel_cost = INF;

    for (i, _) in pairwise_dist.iter().enumerate().take(k) {
        min_travel_cost = std::cmp::min(min_travel_cost, dp[(1 << k) - 1][i] + pairwise_dist[i][k]);
    }

    writeln!(writer, "{}", min_travel_cost).ok();
}

File: beecrowd/1947 - Rota do Taxista.cpp
#include <bits/stdc++.h>

using namespace std;

constexpr int N = 1e4 + 5, K = 15, INF = 1e9;

vector<pair<int, int>> adj[N];
int dp[1 << K][K], pairwiseDist[K + 1][K + 1]; 
int n, m, k;

vector<int> dijkstra(int src) {
    using pq_type = pair<int, int>;

    vector<int> dist(n, INF);
    priority_queue<pq_type, vector<pq_type>, greater<pq_type>> pq;

    dist[src] = 0;
    pq.emplace(0, src);

    while(!pq.empty()) {
        auto [cost, v] = pq.top();

        pq.pop();

        if(cost != dist[v]) {
            continue;
        }

        for(const auto& [u, w] : adj[v]) {
            int total_cost = cost + w;

            if(dist[u] > total_cost) {
                dist[u] = total_cost;
                pq.emplace(total_cost, u);
            }
        }
    }

    return dist;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m >> k;
    
    for(int i = 0; i < m; ++i) {
        int u, v, w;

        cin >> u >> v >> w;

        --u, --v;

        adj[u].emplace_back(v, w);
        adj[v].emplace_back(u, w);
    }

    vector<pair<int, int>> travels(k + 1);

    for(int i = 0; i < k; ++i) {
        int src, dest;

        cin >> src >> dest;

        travels[i] = {src - 1, dest - 1};
    }

    travels[k] = {0, 0};

    for(int i = 0; i <= k; ++i) {
        auto [src1, dest1] = travels[i];

        pairwiseDist[i][i] = dijkstra(src1)[dest1];

        auto costs = dijkstra(dest1);

        for(int j = 0; j <= k; ++j) {
            if(i == j)
                continue;

            int src2 = travels[j].first;

            pairwiseDist[i][j] = costs[src2];
        }
    }

    for(int i = 0; i < (1 << k); ++i)
        for(int j = 0; j < k; ++j)
            dp[i][j] = INF;

    for(int i = 0; i < k; ++i)
        dp[1 << i][i] = pairwiseDist[k][i] + pairwiseDist[i][i];

    for(int i = 0; i < (1 << k); ++i) {
        for(int j = i; j; j -= j & -j) {
            int u = __builtin_ctz(j);

            for(int l = i; l; l -= l & -l) {
                int v = __builtin_ctz(l);

                dp[i][v] = min(dp[i][v], dp[i ^ (1 << v)][u] + pairwiseDist[u][v] + pairwiseDist[v][v]);
            }
        }
    }

    int min_travel_cost {INF};

    for(int i = 0; i < k; ++i)
        min_travel_cost = min(min_travel_cost, dp[(1 << k) - 1][i] + pairwiseDist[i][k]);

    cout << min_travel_cost << '\n';

    return 0;
}